docs: Non-UI API docs rewrite (pt 1) (#8825)
unoplatform/uno: doc/articles/features/file-management.md

# File Management

> [!TIP]
> This article covers Uno-specific information for file management. For a full description of the feature and instructions on using it, consult the UWP documentation: https://docs.microsoft.com/en-us/windows/uwp/files/

 * File management allows shared reading and writing of files across all Uno Platform targets

## Supported features

| Feature        |  Windows  | Android |  iOS  |  Web (WASM)  | macOS | Linux (Skia)  | Win 7 (Skia) | 
|---------------|-------|-------|-------|-------|-------|-------|-|
| `StorageFile` | ✔ | ✔ | ✔| ✔ | ✔| ✔ |✔ |
| `StorageFolder` | ✔ | ✔ | ✔| ✔ | ✔| ✔ |✔ |
| `ApplicationData.Current.LocalFolder` | ✔ | ✔ | ✔| ✔ | ✔| ✔ |✔ |
| `ApplicationData.Current.RoamingFolder` | ✔ | ✔ | ✔| ✔ | ✔| ✔ |✔ |
| `CachedFileManager` | ✔ | partial | partial | partial | partial | partial | partial |

## Overview

Uno supports some of the APIs from the `Windows.Storage` namespace, such as `Windows.Storage.StorageFile` and `Windows.Storage.StorageFolder` for all platforms.

> Both `Windows.Storage` and `System.IO` APIs are available, with some platform specifics defined below. In general, it is best to use `Windows.Storage` APIs when available, as their asynchronous nature allows for transparent interactions with the underlying file system implementations. In addition, `System.IO` cannot work with files which are not owned by the application directly (e.g. files picked by a dialog).

Reviewer: ```suggestion
Both `Windows.Storage` and `System.IO` APIs are available, with some platform specifics defined below. In general, it is best to use `Windows.Storage` APIs when available, as their asynchronous nature allows for transparent interactions with the underlying file system implementations. In addition, `System.IO` cannot work with files that are not owned by the application directly (e.g. files picked by a dialog).
```

Note that for file and folder metadata only `BasicProperties` are partially supported for now. 

> `FileAttributes` and all "advanced properties" (`StorageItemContentProperties`) related to the content of the file, including thumbnail, are not yet supported.

Reviewer: ```suggestion
`FileAttributes` and all "advanced properties" (`StorageItemContentProperties`) related to the content of the file, including the thumbnail, are not yet supported.
```

Note that you can view the content of the **IndexedDB** in the Application tab of your browser, in the **Storage / IndexedDB** section.

## Support for `StorageFile.GetFileFromApplicationUriAsync`

Uno supports the ability to get package files using the [`StorageFile.GetFileFromApplicationUriAsync`](https://docs.microsoft.com/en-us/uwp/api/windows.storage.storagefile.getfilefromapplicationuriasync).

Support per platform may vary:
- On non-WebAssembly targets, the file is available directly as it is a part of the installed package.

> - On WebAssembly, the requested file is part of the application package on the remote server, and is downloaded on demand to avoid increasing the initial application payload size. After it is requested for the first time, the file is then stored in the browser IndexedDB.

Reviewer: ```suggestion
- On WebAssembly, the requested file is part of the application package on the remote server and is downloaded on demand to avoid increasing the initial application payload size. After it is requested for the first time, the file is then stored in the browser IndexedDB.
```

</ItemGroup>
```

## Support for `RandomAccessStreamReference.CreateFromUri`

> Uno Platform supports to creation of a `RandomAccessStreamReference` from an `Uri` (`RandomAccessStreamReference.CreateFromUri`), but note that on WASM downloading a file from a server often causes issues with [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS). 

Reviewer: ```suggestion
Uno Platform supports the creation of a `RandomAccessStreamReference` from an `Uri` (`RandomAccessStreamReference.CreateFromUri`), but note that on WASM downloading a file from a server often causes issues with [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS). 
```

Make sure the server that hosts the file is configured accordingly.

## Support for `CachedFileManager`

> For all targets except for UWP/WinUI and WebAssembly, the `CachedFileManager` does not provide any functionality and its methods immediately return. This allows to easily write code which requires deferring updates on UWP and sharing it across all targets.

Reviewer: ```suggestion
For all targets except for UWP/WinUI and WebAssembly, the `CachedFileManager` does not provide any functionality and its methods immediately return. This allows us to easily write code that requires deferring updates on UWP and sharing it across all targets.
```

> In case of WebAssembly, the behavior of `CachedFileManager` depends on whether the app uses the **File System Access API** or **Download picker**. This is described in detail in [file pickers documentation](windows-storage-pickers.md#webassembly).

Reviewer: ```suggestion
In the case of WebAssembly, the behavior of `CachedFileManager` depends on whether the app uses the **File System Access API** or **Download picker**. This is described in detail in [file pickers documentation](windows-storage-pickers.md#webassembly).
```
